server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/routes/ForwardRoutes.java

```diff
@@ -129,15 +126,10 @@ public Object throwUnknownPath(Request request, Response response) {
             message = "Internal server error - Something went bad on the server side.")
     })
     public Set<String> listForwards(Request request, Response response) throws RecipientRewriteTableException {
-        return Optional.ofNullable(recipientRewriteTable.getAllMappings())
-            .map(mappings ->
-                mappings.entrySet().stream()
-                    .filter(e -> e.getValue().contains(Mapping.Type.Forward))
-                    .map(Map.Entry::getKey)
-                    .flatMap(source -> OptionalUtils.toStream(source.asMailAddress()))
-                    .map(MailAddress::asString)
-                    .collect(Guavate.toImmutableSortedSet()))
-            .orElse(ImmutableSortedSet.of());
+        return recipientRewriteTable.getSourcesForType(Mapping.Type.Forward).stream()
+            .flatMap(source -> OptionalUtils.toStream(source.asMailAddress()))
```

Comment:
IMO you can put those pipeline chain of operations into `recipientRewriteTable.getSourcesForType()`, that also helps to simplify  `ForwardRoutes.listForwards()`, `GroupsRoutes.listGroups()` and `AliasRoutes.listAliasesOfAddress()`

Reply:
API  strong-typing is more important than local code optimisation

 1. RRT::getSourcesForType could be returning a stream
 2. Maybe it means that working with MappingSources is too hard and we might need a `Stream<String> asMailaddressString()` method (but I am not sure about that point. OR we could be teaching Jackson how to be writing a MappingSource directly. A `JsonTransformerModule` for MappingSources seems to be the most clever way to handle it...

Reply:
I tried to go through a `JsonTransformerModule` at the end and needed to implement some `compareTo` method for MappingSource to sort it in the desirable way. Let me know if we can do better !